lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -1105,3 +1119,125 @@ gDecodeAddress decodeShelley text =
+
+-- not delegating, joined as it follows from discovered delegation certificates
+-- and effective joining epoch and its start time is specified
+delegating
+    :: ApiT PoolId
+    -> ApiWalletDelegation
+delegating pId =
+    ApiWalletDelegation
+    (ApiDelegationStatus $ ApiT $ Delegating pId)
+    Nothing
+
+-- delegating, quitted as it follows from discovered delegation certificates
+-- and effective quitting epoch and its start time is specified
+delegatingAboutToQuit
+    :: ApiT PoolId
+    -> EpochNo
+    -> SlotParameters
+    -> ApiWalletDelegation
+delegatingAboutToQuit pId epochNo sp =
+    ApiWalletDelegation
+    (ApiDelegationStatus $ ApiT $ Delegating pId)
+    (Just $
+        ApiWalletDelegationNext
+        ( ApiDelegationStatus $ ApiT NotDelegating)
+        ( ApiEpochInfo
+            (ApiT $ epochNo + 2) (W.epochStartTime sp (epochNo + 2))
+        )
+    )
+
+-- delegating, rejoined as it follows from discovered delegation certificates
+-- and effective rejoining epoch to another stake pool and its start time is specified
+delegatingAboutToRejoin
+    :: ApiT PoolId
+    -> ApiT PoolId
+    -> EpochNo
+    -> SlotParameters
+    -> ApiWalletDelegation
+delegatingAboutToRejoin pId1 pId2 epochNo sp =
+    ApiWalletDelegation
+    (ApiDelegationStatus $ ApiT $ Delegating pId1)
+    (Just $
+        ApiWalletDelegationNext
+        ( ApiDelegationStatus $ ApiT $ Delegating pId2)
+        ( ApiEpochInfo
+            (ApiT $ epochNo + 2) (W.epochStartTime sp (epochNo + 2))
+        )
+    )
+
+newtype ErrDelegationsDiscoveredInconsistency
+    = ErrDelegationsDiscoveredInconsistency Text
+    deriving (Show, Eq)
+
+toApiWalletDelegation
+    :: EpochNo
+    -> SlotParameters
+    -> [DelegationDiscovered]
+    -> Either ErrDelegationsDiscoveredInconsistency ApiWalletDelegation
+toApiWalletDelegation currentEpochNo sp dlgs = case dlgs of
+    [] ->
+        Right notDelegating
+    [(DelegationDiscovered (SlotId epochN _) poolIdMaybe)] ->
+        case poolIdMaybe of
+            Nothing ->
+                err "there should not be quiting delegation when \
+                \no prior joining delegation is present"
+            Just pId ->
+                if currentEpochNo < epochN + 2 then
+                    Right $ notDelegatingAboutToJoin (ApiT pId) epochN sp
+                else
+                    Right $ delegating (ApiT pId)
+    [ (DelegationDiscovered (SlotId epochN1 _) poolIdMaybe1)
+        , (DelegationDiscovered (SlotId epochN2 _) poolIdMaybe2)] ->
+        if epochN2 > epochN1 then
+            err "delegation discovered later cannot have epoch number \
+                \older than the previously discovered delegation"
+        else case (poolIdMaybe1, poolIdMaybe2) of
+            (Nothing, Nothing) ->
+                err "there should not be two quit delegation certificates\
+                      \ discovered in row"
+            (Just pId, Nothing) ->
+                if currentEpochNo < epochN1 + 2 then
```

Comment:
Idea: maybe we could remove the duplication of this `if currentEpochNo < epochN1 + 2` by first:
```haskell
let activeDelegation = safeLast $ filter (...) dlgs
let upcomingDelegation = safeLast $ filter (...) dlgs
case (activeDelegation, upcomingDelegation) of
...
```

I came to realise that this function is pretty complex and I hadn't checked how it works.

Reply:
yeah, I was contemplating this several times. For me explicit if ... else clauses are the clearest to understand in this situation